subprojects/model-core/src/main/java/org/gradle/internal/reflect/annotations/impl/DefaultTypeAnnotationMetadataStore.java

```diff
@@ -349,6 +359,24 @@ private void processMethodAnnotations(Method method, Map<String, PropertyAnnotat
         }
     }
 
+    private static void validateSetterForMutableType(PropertyAccessorType accessorType, Method method, ValidationErrorsBuilder errorsBuilder) {
+        Type setterType = getRawSetterType(accessorType, method);
+        if (MUTABLE_NON_FINAL_TYPES.contains(setterType)) {
```

Comment:
I think we should check for assignability here, not exact match, so we'd catch `RegularFileProperty` etc. as well. We should add tests for subtypes, too.

Reply:
Sure, that was one of my questions in a buried comment here.
Which seems more logical to you:
* gather all supertypes of the iterated class and do a `containsAny`
* iterate through the blacklist and do an isAssignableFrom?

Reply:
is it guaranteed that subclasses of these will have corresponding setters for their newly introduced fields?

Reply:
These are types that should only be extended by Gradle, so we control all the implementations. For those implementations we can be sure that setters are a bad idea.

Since there are two types that we are checking in the only known usage of this class, I'd say let's iterate the blacklist.

Reply:
Should we warn for other known mutable types, like arrays or Collections and stuff?

Reply:
Let's focus on these for now. The current plan is to push people towards using these types exclusively anyway (via gentle nudges in future additional validations), so other types will become less important.